pkg/bundle/sync.go

```diff
@@ -145,89 +150,163 @@ func (b *bundle) secretBundle(ctx context.Context, ref *trustapi.SourceObjectKey
+			needsPatch = true
+		}
 
-	// Match, return do nothing
-	if cmdata, ok := configMap.Data[target.ConfigMap.Key]; !ok || cmdata != data {
-		if configMap.Data == nil {
-			configMap.Data = make(map[string]string)
+		// Check if we need to migrate the ConfigMap managed fields to the Apply field operation
+		needsMigration := false
+		for _, mf := range configMap.ManagedFields {
+			if mf.Manager == fieldManager && mf.Operation == metav1.ManagedFieldsOperationUpdate {
+				needsMigration = true
+				needsPatch = true
+			}
+		}
+
+		if needsMigration {
+			if err := b.migrateToApply(ctx, configMap); err != nil {
+				return false, fmt.Errorf("failed to migrate ConfigMap %s/%s to Apply: %w", namespace, name, err)
+			}
 		}
-		configMap.Data[target.ConfigMap.Key] = data
-		needsUpdate = true
 	}
 
 	// Exit early if no update is needed
-	if !needsUpdate {
+	if !needsPatch {
 		return false, nil
 	}
 
-	if err := b.directClient.Update(ctx, &configMap); err != nil {
-		return true, fmt.Errorf("failed to update configmap %s/%s with bundle: %w", namespace, bundle.Name, err)
+	configMapPatch := coreapplyconfig.
+		ConfigMap(name, namespace).
+		WithOwnerReferences(
+			v1.OwnerReference().
+				WithAPIVersion(trustapi.SchemeGroupVersion.String()).
+				WithKind(trustapi.BundleKind).
+				WithName(bundle.GetName()).
+				WithUID(bundle.GetUID()).
+				WithBlockOwnerDeletion(true).
+				WithController(true),
+		).
+		WithData(map[string]string{target.ConfigMap.Key: data})
+
+	if err = b.patchResource(ctx, configMapPatch); err != nil {
+		return false, fmt.Errorf("failed to patch configMap %s/%s: %w", namespace, bundle.Name, err)
 	}
 
 	log.V(2).Info("synced bundle to namespace")
 
 	return true, nil
 }
+
+func (b *bundle) patchResource(ctx context.Context, obj interface{}) error {
+	if b.patchResourceOverwrite != nil {
+		return b.patchResourceOverwrite(ctx, obj)
+	}
+
+	applyConfig, ok := obj.(*coreapplyconfig.ConfigMapApplyConfiguration)
+	if !ok {
+		return fmt.Errorf("expected *coreapplyconfig.ConfigMapApplyConfiguration, got %T", obj)
+	}
+
+	configMap, patch, err := ssa_client.GenerateConfigMapPatch(applyConfig)
+	if err != nil {
+		return fmt.Errorf("failed to generate patch: %w", err)
+	}
+
+	err = b.directClient.Patch(ctx, configMap, patch, &client.SubResourcePatchOptions{
+		PatchOptions: client.PatchOptions{
+			FieldManager: fieldManager,
+			Force:        pointer.Bool(true),
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	// If the configMap is empty, delete it
+	if len(configMap.Data) == 0 {
+		return b.directClient.Delete(ctx, configMap)
+	}
+
+	return nil
+}
+
+// MIGRATION: This is a migration function that migrates the ownership of
+// fields from the Update operation to the Apply operation. This is required
+// to ensure that the apply operations will also remove fields that were
+// created by the Update operation.
+func (b *bundle) migrateToApply(ctx context.Context, obj client.Object) error {
```

Comment:
TODO: this function still needs a test

Reply:
@irbekrm I added a fix here, could you test if this works?
I haven't added a test yet, if you or someone else want to try to add one, feel free to do so.

Reply:
Thanks @inteon I tested it this morning with the same strategy as in https://github.com/cert-manager/trust-manager/pull/89#issuecomment-1378452644 - I still see the namespace selector remain on Bundle status after it has been removed from the spec.

FWIW, I think you cannot test it by deploying trust-manager using the make commands as they uninstall every time before reinstalling.
I also had to add 'update' for configmaps to trust's clusterrole to get it working.

Reply:
@irbekrm I was only migrating the ConfigMap resource. I added a migration for the bundle resource too now.